autopilot/betweenness_centrality.go

```diff
@@ -0,0 +1,177 @@
+// to measure control of individual nodes over the whole network.
+type BetweennessCentrality struct {
+	centrality map[NodeID]float64
+}
+
+// NewBetweennessCentralityMetric creates a new BetweennessCentrality instance.
+func NewBetweennessCentralityMetric() *BetweennessCentrality {
+	return &BetweennessCentrality{}
+}
+
+// Name returns the name of the metric.
+func (bc *BetweennessCentrality) Name() string {
+	return "betweeness_centrality"
+}
+
+// betweennessCentrality is the core of Brandes' algorithm.
+// We first calculate the shortest paths from the start node s to all other
+// nodes with BFS, then update the betweenness centrality values by using
+// Brandes' dependency trick.
+// For detailed explanation please read:
+// https://www.cl.cam.ac.uk/teaching/1617/MLRD/handbook/brandes.html
+func betweennessCentrality(g *SimpleGraph, s int, centrality []float64) {
+	// Pred[w] is the list of nodes that immediately precede w on a
+	// shortest path from s to t for each node t.
+	Pred := make([][]int, len(g.Nodes))
+
+	// sigma[t] is the number of shortest paths between nodes s and t for
+	// each node t.
+	sigma := make([]int, len(g.Nodes))
+	sigma[s] = 1
+
+	// dist[t] holds the distance between s and t for each node t. We initialize
+	// this to -1 (meaning infinity) for each t != s.
+	dist := make([]int, len(g.Nodes))
+	for i := range dist {
+		dist[i] = -1
+	}
+
+	dist[s] = 0
+
+	var (
+		S stack
+		Q queue
+	)
+	Q.Push(s)
+
+	// BFS to calculate the shortest paths (sigma and Pred)
+	// from s to t for each node t.
+	for !Q.Empty() {
+		v := Q.Front()
+		Q.Pop()
+		S.Push(v)
+
+		for _, w := range g.Adj[v] {
+			// If distance from s to w is infinity (-1)
+			// then set it and enqueue w.
+			if dist[w] < 0 {
+				dist[w] = dist[v] + 1
+				Q.Push(w)
+			}
+
+			// If w is on a shortest path the update
+			// sigma and add v to w's predecessor list.
+			if dist[w] == dist[v]+1 {
+				sigma[w] += sigma[v]
+				Pred[w] = append(Pred[w], v)
+			}
+		}
+	}
+
+	// delta[v] is the ratio of the shortest paths between s and t that go
+	// through v and the total number of shortest paths between s and t.
+	// If we have delta then the betweenness centrality is simply the sum
+	// of delta[w] for each w != s.
+	delta := make([]float64, len(g.Nodes))
+
+	for !S.Empty() {
+		w := S.Top()
+		S.Pop()
+
+		// Pred[w] is the list of nodes that immediately precede w on a
+		// shortest path from s.
+		for _, v := range Pred[w] {
+			// Update delta using Brandes' equation.
+			delta[v] += (float64(sigma[v]) / float64(sigma[w])) * (1.0 + delta[w])
+		}
+
+		if w != s {
+			// As noted above centrality is simply the sum
+			// of delta[w] for each w != s.
+			centrality[w] += delta[w]
+		}
+	}
+}
+
+// Refresh recaculates and stores centrality values.
+func (bc *BetweennessCentrality) Refresh(graph ChannelGraph) error {
+	cache, err := NewSimpleGraph(graph)
+	if err != nil {
+		return err
+	}
+
+	// TODO: parallelize updates to centrality.
```

Comment:
should be pretty straight forward to launch this in `N` go routines?

Reply:
added a new commit to do this, ptal